examples/all-clusters-app/all-clusters-common/src/EnergyEvseDelegateImpl.cpp

```diff
@@ -0,0 +1,771 @@
+{
+    ChipLogProgress(AppServer, "EnergyEvseDelegate::Disable()");
+
+    /* Update State */
+    switch (mHwState)
+    {
+    case StateEnum::kNotPluggedIn:
+        SetState(StateEnum::kNotPluggedIn);
+        break;
+
+    case StateEnum::kPluggedInNoDemand:
+        SetState(StateEnum::kPluggedInNoDemand);
+        break;
+
+    case StateEnum::kPluggedInDemand:
+        SetState(StateEnum::kPluggedInDemand);
+        break;
+
+    default:
+        ChipLogError(AppServer, "Unexpected EVSE hardware state\n");
+        SetState(StateEnum::kFault);
+        break;
+    }
+
+    /* update SupplyState */
+    SetSupplyState(SupplyStateEnum::kDisabled);
+
+    /* update ChargingEnabledUntil & DischargingEnabledUntil to show 0 */
+    SetChargingEnabledUntil(0);
+    SetDischargingEnabledUntil(0);
+
+    /* update MinimumChargeCurrent & MaximumChargeCurrent to 0 */
+    SetMinimumChargeCurrent(0);
+    SetMaximumChargeCurrent(0);
+
+    /* update MaximumDischargeCurrent to 0 */
+    SetMaximumDischargeCurrent(0);
+
+    // TODO: Generate events
+
+    return Status::Success;
+}
+
+/**
+ * @brief   Called when EVSE cluster receives EnableCharging command
+ *
+ * @param chargingEnabledUntil (can be null to indefinite charging)
+ * @param minimumChargeCurrent (in mA)
+ * @param maximumChargeCurrent (in mA)
+ */
+Status EnergyEvseDelegate::EnableCharging(const DataModel::Nullable<uint32_t> & chargingEnabledUntil,
+                                          const int64_t & minimumChargeCurrent, const int64_t & maximumChargeCurrent)
+{
+    ChipLogProgress(AppServer, "EnergyEvseDelegate::EnableCharging()");
+
+    if (maximumChargeCurrent < kMinimumChargeCurrent || maximumChargeCurrent > kMaximumChargeCurrent)
+    {
+        ChipLogError(NotSpecified, "Maximum Current outside limits");
+        return Status::ConstraintError;
+    }
+
+    if (minimumChargeCurrent < kMinimumChargeCurrent || minimumChargeCurrent > kMaximumChargeCurrent)
+    {
+        ChipLogError(NotSpecified, "Maximum Current outside limits");
+        return Status::ConstraintError;
+    }
+
+    if (minimumChargeCurrent > maximumChargeCurrent)
+    {
+        ChipLogError(NotSpecified, "Minium Current > Maximum Current!");
+        return Status::ConstraintError;
+    }
+
+    if (chargingEnabledUntil.IsNull())
+    {
+        /* Charging enabled indefinitely */
+        ChipLogError(NotSpecified, "Charging enabled indefinitely");
+    }
+    else
+    {
+        /* check chargingEnabledUntil is in the future */
+        ChipLogError(NotSpecified, "Charging enabled until: %lu", static_cast<long unsigned int>(chargingEnabledUntil.Value()));
+        // TODO
+        // if (checkChargingEnabled)
+    }
+
+    /* Check current state isn't already enabled */
+
+    /* If charging is already enabled, check that the parameters may have
+       changed, these may override an existing charging command */
+    switch (mHwState)
+    {
+    case StateEnum::kNotPluggedIn:
```

Comment:
would it be shorter to have:

```
switch (mHwState) {
   case StateEnum::kNotPluggedIn:
   case StateEnum::kPluggedInNoDemand:
   case StateEnum::kPluggedInDemand:
        // explain why you are calling setState, which seems to just set mHWState
        // again with the mHWState value
        if (SetState(mHWState) != Status::Success) {
           // handle
        }
        break;
    default:
        ChipLogError(AppServer, "Unexpected EVSE hardware state");
        SetState(StateEnum::kFault);
        break;
}
```

Notable changes beside shortness of code:
   - please handle things returning statues
   - explain why you are calling SetState with the current state
   - do not append a "\n" for ChipLogError calls.

Reply:
I just noticed that kPluggedInDemand is different. The way this was written made it non-obvious. That case should obviously be special-cased. Still need error handling if error is possible.

Reply:
This delegate code is still being developed now - so is incomplete, and I've been trying to get the basic all clusters landed first then update the delegate in a new PR.

The actual handling of the different states in the delegate is more complex and ONLY uses 3 possible states from the ones that the cluster can report.
i.e. the hardware can only set the 3 states, but the charging aspect of the enum are based on the application client enabling or disabling charging.

So the switch statement will be more complex when finished.

I'll ensure that the comments above are fed into that.

I've just raised a new Issue to 'complete' the delegate code...
https://github.com/project-chip/connectedhomeip/issues/30997